lib/charms/dex_auth/v0/dex_oidc_config.py

```diff
@@ -0,0 +1,372 @@
+        charm (CharmBase): the provider application
+        refresh_event: (list, optional): list of BoundEvents that this manager should handle.
+                       Use this to update the data sent on this relation on demand.
+        relation_name (str, optional): the name of the relation
+
+    Attributes:
+        charm (CharmBase): variable for storing the requirer application
+        relation_name (str): variable for storing the name of the relation
+    """
+
+    on = DexOidcConfigEvents()
+
+    def __init__(
+        self,
+        charm: CharmBase,
+        refresh_event: Optional[Union[BoundEvent, List[BoundEvent]]] = None,
+        relation_name: Optional[str] = DEFAULT_RELATION_NAME,
+    ):
+        super().__init__(charm, relation_name)
+        self._charm = charm
+        self._relation_name = relation_name
+        self._requirer_wrapper = DexOidcConfigRequirerWrapper(self._charm, self._relation_name)
+
+        self.framework.observe(
+            self._charm.on[self._relation_name].relation_changed, self._on_relation_changed
+        )
+
+        self.framework.observe(
+            self._charm.on[self._relation_name].relation_broken, self._on_relation_broken
+        )
+
+        if refresh_event:
+            if not isinstance(refresh_event, (tuple, list)):
+                refresh_event = [refresh_event]
+            for evt in refresh_event:
+                self.framework.observe(evt, self._on_relation_changed)
+
+    def get_data(self) -> DexOidcConfigObject:
+        """Return a DexOidcConfigObject."""
+        return self._requirer_wrapper.get_data()
+
+    def _on_relation_changed(self, event: BoundEvent) -> None:
+        """Handle relation-changed event for this relation."""
+        self.on.updated.emit(event.relation)
+
+    def _on_relation_broken(self, event: BoundEvent) -> None:
+        """Handle relation-broken event for this relation."""
+        self.on.updated.emit(event.relation)
+
+
+class DexOidcConfigRequirerWrapper(Object):
+    """Wrapper for the relation data getting logic.
+
+    Args:
+        charm (CharmBase): the requirer application
+        relation_name (str, optional): the name of the relation
+
+    Attributes:
+        relation_name (str): variable for storing the name of the relation
+    """
+
+    def __init__(self, charm, relation_name: Optional[str] = DEFAULT_RELATION_NAME):
+        super().__init__(charm, relation_name)
+        self.relation_name = relation_name
+
+    @staticmethod
+    def _validate_relation(relation: Relation) -> None:
+        """Series of checks for the relation and relation data.
+
+        Args:
+            relation (Relation): the relation object to run the checks on
+
+        Raises:
+            DexOidcConfigRelationDataMissingError if data is missing or incomplete
+            DexOidcConfigRelationMissingError: if there is no related application
+        """
+        # Raise if there is no related application
+        if not relation:
```

Comment:
> If I'm correct, the relation object has not been "empty", so this should be if relation is None or not?

I am not sure what do you mean by empty, do you mind providing more info?

> If yes, can we unite also type hinting to be Optional[Relation].

It is not optional though, the purpose of this validation is to always raise if a requirer wants to get data from an application that it's not even related. We are not doing this check anywhere else, not even when the requirer object is instantiated.

Reply:
> I am not sure what do you mean by empty, do you mind providing more info?

Checking the [source code](https://github.com/canonical/operator/blob/5a21cd28a47b8d5f1142d794130ae70f23b5cd26/ops/model.py#L1548) of Relation object, I do not see how this statement could be True. There is no `__eq__` function defined, so that's why I think that `if not relation:` could be True only if `relation is None`.

```python
>>> class A:
...  def __init__(self, a: str, b: str, c: str):
...   self.a = a
...   self.b = b
...   self.c = c
...  def __repr__(self):
...   return f"a:{self.a} b:{self.b} c:{self.c}"
... 
>>> a = A("1", "2", "3")
>>> if a:
...  print(a)
... 
a:1 b:2 c:3
```

Reply:
`relation` is a `Relation` object returned by [`get_relation`](https://ops.readthedocs.io/en/latest/index.html#ops.Model.get_relation). The latter will return `None` if the relation is not established. What we are checking here is that there is in fact a relation and raise when it is `None`.

That being said, I will modify the docstrings to better define what this arg is for.

Reply:
Can we also change the type hints?
e.g.
```python
def _validate_relation(relation: Optional[Relation]) -> None:
```